pypose/module/ipddp.py

```diff
@@ -0,0 +1,573 @@
+                        # relax a bit for numerical stability, strange; todo: any for each sample in a batch?
+                        failed=True
+                        continue
+                    else:
+                        idx = torch.all(candidate<=self.filter,-1)
+                        self.filter = self.filter[torch.logical_not(idx)]
+                        if self.filter.ndim <= 2:
+                            self.filter = self.filter.unsqueeze(0)
+                        self.filter=torch.cat((self.filter, candidate.unsqueeze(-2)), dim=-2)
+                        break
+
+        if (failed):
+            self.stepsize, self.failed= 0.0, failed
+        else:
+            self.x, self.u, self.y, self.s, self.c = xnew, unew, ynew, snew, cnew
+            self.cost, self.err, self.stepsize, self.step, self.failed = cost, err, stepsize, step, False
+
+    def solver(self, verbose=False):
+        r'''
+        Call forwardpass and backwardpass to solve an optimal trajectory for general nonlinear system
+        '''
+        time_start = time.time()
+
+        for t in range(self.T):
+            self.x[...,t+1,:], _ = self.f_fn(self.x[...,t,:],self.u[...,t,:])
+        self.c = self.c_fn(self.x[...,:-1,:], self.u)
+        self.cost = self.q_fn(self.x[...,:-1,:], self.u).sum(-1) \
+                    + self.p_fn(self.x[...,-1,:],torch.zeros_like(self.u[...,-1,:])).sum(-1)
+        self.mu = self.cost / self.T / self.s[...,0,:].shape[-1]
+        self.resetfilter()
+        self.reg, self.bp_failed, self.recovery = 0.0, False, 0
+
+        for iter in range(self.maxiter):
+            while True:
+                self.backwardpasscompact()
+                if not self.bp_failed:
+                    break
+
+            self.forwardpasscompact()
+            time_used = time.time() - time_start
+            if verbose:
+                if (iter % 5 == 1):
+                    print('\n')
+                    print('Iteration','Time','mu','Cost','Opt. error','Reg. power','Stepsize')
+                    print('\n')
+                print('%-12d%-12.4g%-12.4g%-12.4g%-12.4g%-12d%-12.3f\n'%(
+                            iter, time_used, self.mu, self.cost, self.opterr, self.reg, self.stepsize))
+
+            #-----------termination conditions---------------
+            if (max(self.opterr, self.mu)<=self.tol):
+                print("~~~Optimality reached~~~")
+                break
+
+            if (self.opterr <= 0.2*self.mu):
+                self.mu = max(self.tol/10.0, min(0.2*self.mu, pow(self.mu, 1.2) ) )
+                self.resetfilter()
+                self.reg, self.bp_failed, self.recovery = 0.0, False, 0
+
+            if iter == self.maxiter - 1:
+                print("max iter", self.maxiter, "reached, not the optimal one!")
+            #-----------------------------------------------
+        return self
+
+    def forward(self, fp_list):
```

Comment:
I'm kind of confused about solver and forward. It seems in the example, forward is used. In the test case, solver is used. I suggest we keep the consistency with the lqr and mpc with a forward would be great. Also, try to keep similar arg list of the forward function.

Reply:
In the test, it only cares about solving the traj, and does not create a computational graph for learning; in the training example, it first calls solver() to solve the trajectory, then use forward to create the computational graph for autodiff.